include/mata/nfa.hh

```diff
@@ -580,29 +580,21 @@ public:
         }
     }
 
-    //TODO: why this? Maybe we could have add_state(int state)
-    //Btw, why do we need adding states actually, we could just add states with transitions, or with initial/final states.
-    void increase_size(size_t size)
-    {
-        assert(this->delta.post_size() <= size);
-        delta.increase_size(size);
-    }
-
-    /**
-     * Increase size to include @p state.
-     * @param state[in] The new state to be included.
-     */
-    void increase_size_for_state(const State state)
-    {
-        increase_size(state + 1);
-    }
-
     /**
      * Add a new state to the automaton.
      * @return The newly created state.
      */
     State add_state();
 
+    State add_state(State state)
+    {
+        if (state < size())
+            return size() - 1;
```

Comment:
The semantic is quite strange, but maybe it is just me. You add state, and you return the maximal number of states? I would expected add_state to return nothing.

Reply:
But adding states does not fail, or does it? I know of some usages such as this (i.e. adding element returns maximal element or  size of the states). I'm just thinking the name of the function could be a little bit confusing. I think maybe docstring here would be enough, i.e. `@returns: size of the state space`.

Reply:
The problem here is that according to the previous semantics, adding new states would fail when the state to be added was already in the automaton. That is, the state to be added had internal value less than the number of states. In such a case, an exception would be thrown.

We do not want to throw exceptions these days, so we need to let the user know in other ways that the state was not added. We could return the value of the state no matter what is the outcome, but that would eliminate the option for the user to detect when they try to add state already in the automaton.

Furthermore, the function does not return the number of states. It returns the internal value oof the newly added state. Otherwise, if the state was already present in the automaton, it should warn the user that the state was not added (it failed in adding the state). This would be the case where `std::optional` should return null value.

The name is confusing because, semantically, there are two different operations bundled within a single function. One, adding a new state. That is it. There it makes sense to return the internal value of the newly added state. Two, increasing size of an automaton to a passed value. This operation can fail if you try to increase the size of the automaton to a size smaller than the actual size.

Reply:
The problem is that `resize` on `Nfa` no longer makes sense since now `delta` has size but `Nfa` does not. I would not go for `optional`. It's introduction of a new abstraction to library which may complicate things for some programmers. I would use it to solve it some crucial problem, not this one which could be resolved by adding comment.

Reply:
I agree that returning `-1` is not the best idea since it is still valid value of state.

Reply:
> I agree that returning `-1` is not the best idea since it is still valid value of state.

When we succeed in adding a new state, we return its number. When we fail (as explained above), we have to return something. That something is currently the last state in an automaton (not the one asked to be added, but the last one). If we failed, we cannot return `size()`, otherwise it would seem as if the state was added successfully. This value would be the internal value of a new state, but it is not a new state, in fact. We can only return values higher than the maximal state or higher for the function to make sense and allow the user to check whether the addition succeeded.

Reply:
Also it does not make sense to talk about `resizing` or whatever in context of `Nfa`. There is nothing to resize there. We just need to agree on 1) whether we should return something or not, 2) if we want return something then what?

Reply:
What does it mean that we fail to add state? From definition, state set of automaton is just a set. You can always add to set, there is no concept of failed adding to set in mathematics.

Reply:
> Also it does not make sense to talk about `resizing` or whatever in context of `Nfa`. There is nothing to resize there. We just need to agree on 1) whether we should return something or not, 2) if we want return something then what?

This was the previous semantics of the method `Nfa::increase_size(size_t size)`. If we are OK with changing the semantics, then I have no issue with always returning the value of the state, for the sake of consistency between `Nfa::add_state(void)` and `Nfa::add_state(State state)`. We need to return the value of a newly added state for `Nfa::add_state(void)` so that the user can use this state afterwards.

> What does it mean that we fail to add state? From definition, state set of automaton is just a set. You can always add to set, there is no concept of failed adding to set in mathematics.

This was the previous semantics of the method `Nfa::increase_size(size_t size)` with sizes less than the current number of states or equal to them. If we agree on `Nfa::add_state(State state)` to follow mathematics definition of a set for our set of states, then again, just return the value of the added state at all times.